testing/test_unittest.py

```diff
@@ -735,3 +735,23 @@ def test_skip(self):
         *SKIP*[1]*skipping due to reasons*
         *1 skipped*
     """)
+
+def test_class_method_containing_test_issue1558(testdir):
+    testdir.makepyfile(test_foo="""
+        import unittest
+
+        class MyTestMeta(type):
+            def __new__(cls, name, bases, dct):
+                for name in ['should_run', 'should_not_run']:
+                    test_name = 'test_' + name
+                    dct[test_name] = lambda self: None
+                    dct[test_name].__name__ = test_name
+
+                dct['test_should_not_run'].__test__ = False
+                return super(MyTestMeta, cls).__new__(cls, name, bases, dct)
+
+        class MyTestCase(unittest.TestCase):
+            __metaclass__ = MyTestMeta
```

Comment:
It would complicate things considerably, but I the solution I posted above should work. :grin: